### PR TITLE
Unecessary asset loading

### PR DESCRIPTION
**Pull request by zoonage**
__STORAGE__ 
- Warn on trying to load an asset with nothing except the storage holding a handle #628.

__ANIMATION__ 
- Adds some profiling code to animation crate that wasn't submitted during #503.

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/642)

### CODE REVIEW

File: amethyst_assets/src/storage.rs [hunk 1]
```diff
@@ -93,6 +95,14 @@ impl<A: Asset> AssetStorage<A> {
     /// Get an asset from a given asset handle.
     pub fn get(&self, handle: &Handle<A>) -> Option<&A> {
         if self.bitset.contains(handle.id()) {
+            // Add a warning if a handle is unique (i.e. asset does not
+            // need to be loaded as it is not used by anything) 
+            // https://github.com/amethyst/amethyst/issues/628
+            if handle.is_unique() {
```
Comment (torkleyy) [inline amethyst_assets/src/storage.rs]: Err this condition will always be false.

**Comment by zoonage** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[amethyst_assets/src/storage.rs, line 101 at r1](https://beta.reviewable.io/reviews/amethyst/amethyst/642#-LA2dN12buLzsUtKL8Ow-r1-101:-LACYlCG06PkWOfsrRrF:b-49qly) ([raw file](https://github.com/amethyst/amethyst/blob/7b62e11c9a14f884a3708d30921f1ad1586fbe6e/amethyst_assets/src/storage.rs#L101)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Err this condition will always be false.
</blockquote></details>

Is that because the `is_unqieu()` method will always be false or because it is in the function? Either way I'm not sure I understand why it would always be false, could you explain?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/amethyst/amethyst/642)*

**Comment by torkleyy** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[amethyst_assets/src/storage.rs, line 101 at r1](https://beta.reviewable.io/reviews/amethyst/amethyst/642#-LA2dN12buLzsUtKL8Ow-r1-101:-LAJS4Vn4qCVluuM8D1O:b2rl09c) ([raw file](https://github.com/amethyst/amethyst/blob/7b62e11c9a14f884a3708d30921f1ad1586fbe6e/amethyst_assets/src/storage.rs#L101)):*
<details><summary><i>Previously, zoonage (Aaron George) wrote…</i></summary><blockquote>

Is that because the `is_unqieu()` method will always be false or because it is in the function? Either way I'm not sure I understand why it would always be false, could you explain?
</blockquote></details>

Sorry, I was in a hurry when I reviewed. `is_unique` cannot return `true` because there is one handle stored in the `AssetStorage` and another one was passed `AssetStorage::get` (this method). Thus, the handle has at least two refs.

This check needs to go into the processing, when the asset got received from the `Loader`. I added another comment to point you to the right place.

---

*[amethyst_assets/src/storage.rs, line 194 at r1](https://beta.reviewable.io/reviews/amethyst/amethyst/642#-LAJSeHa8SwscnFkN3au:-LAJSeHa8SwscnFkN3av:b-8q8xo1) ([raw file](https://github.com/amethyst/amethyst/blob/7b62e11c9a14f884a3708d30921f1ad1586fbe6e/amethyst_assets/src/storage.rs#L194)):*
> ```Rust
>                         }
>                     };
> ```

The check can go about here.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/amethyst/amethyst/642)*

**Comment by jojolepro** [comment]
Are you using the latest version of rustfmt-preview?

**Comment by zoonage** [comment]
This was done with 1.26 nightly, just updated to 1.27

**Comment by zoonage** [comment]
Review status: 0 of 65 files reviewed at latest revision, 2 unresolved discussions.

---

*[amethyst_assets/src/storage.rs, line 194 at r1](https://beta.reviewable.io/reviews/amethyst/amethyst/642#-LAJSeHa8SwscnFkN3au:-LAJZEuf7nImR-4VcBJm:b-x0ktkf) ([raw file](https://github.com/amethyst/amethyst/blob/7b62e11c9a14f884a3708d30921f1ad1586fbe6e/amethyst_assets/src/storage.rs#L194)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

The check can go about here.
</blockquote></details>

Why do we only check for this in `NewAssets`? Wouldn't assets that are being `HotReload` need to be checked as well?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/amethyst/amethyst/642)*

**Comment by Rhuagh** [comment]
We use the stable preview

File: amethyst_assets/src/asset.rs [hunk 2]
```diff
@@ -3,6 +3,12 @@ use std::sync::Arc;
 use amethyst_core::specs::UnprotectedStorage;
 #[cfg(feature = "profiler")]
 #[macro_use]
+#[cfg(feature = "profiler")]
+#[macro_use]
+#[cfg(feature = "profiler")]
+#[macro_use]
+#[cfg(feature = "profiler")]
+#[macro_use]
```
Comment (Xaeroxe) [inline amethyst_assets/src/asset.rs]: Yerr this is almost certainly a mistake

**Comment by zoonage** [comment]
re-run this with the stable toolchain

---

Review status: 0 of 93 files reviewed at latest revision, 3 unresolved discussions.

---

*[amethyst_assets/src/asset.rs, line 11 at r3](https://beta.reviewable.io/reviews/amethyst/amethyst/642#-LAJZk0PneZuPbqqg6C--r3-11:-LAJaKE63_1oIjq7uqtC:bc5j2tg) ([raw file](https://github.com/amethyst/amethyst/blob/dea793684fda874b3e55a61454561019a930ee84/amethyst_assets/src/asset.rs#L11)):*
<details><summary><i>Previously, Xaeroxe (Jacob Kiesel) wrote…</i></summary><blockquote>

Yerr this is almost certainly a mistake
</blockquote></details>

it most definitely is, I've removed these

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/amethyst/amethyst/642#-:-LAJaGnIB9NX3XfZyaiI:bgy3yxh)*

**Comment by zoonage** [comment]
To clean this up I'll reset this branch back to develop and re-add the handle change so that we  can do this without rust fmt and the additional profiling changes (which I question the value of after I did more work on the profiling crate)

File: amethyst_assets/src/storage.rs [hunk 3]
```diff
@@ -148,6 +148,8 @@ impl<A: Asset> AssetStorage<A> {
                     name,
                     tracker,
                 } => {
+
```
Comment (torkleyy) [inline amethyst_assets/src/storage.rs]: Please remove the whitespace and squash the commits, then this is ready to get merged :+1:

File: amethyst_assets/src/storage.rs [hunk 4]
```diff
@@ -147,7 +147,7 @@ impl<A: Asset> AssetStorage<A> {
                     handle,
                     name,
                     tracker,
-                } => {
+                } => {
```
Comment (jojolepro) [inline amethyst_assets/src/storage.rs]: what is this?

Comment (torkleyy) [reply amethyst_assets/src/storage.rs]: extra whitespace supply for hard times ^^ nah, it'll be cleaned up with the next cargo fmt

Comment (torkleyy) [reply amethyst_assets/src/storage.rs]: extra whitespace supply for hard times ^^ nah, it'll be cleaned up with the next cargo fmt

Comment (torkleyy) [reply amethyst_assets/src/storage.rs]: The formatting bot would still be nice - maybe just a bot that creates PRs if there are changes.

**Pull request merged**